File: games/game-example/bots/java/src/core/NetworkingClient.java
package core;

import com.beust.jcommander.JCommander;
import com.google.gson.Gson;
import core.api.Response;
import core.api.InitialData;
import core.api.MatchState;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;


/**
 * Handles the connection to the game engine and takes
 * care of sending and retrieving data.
 **/
public class NetworkingClient extends WebSocketClient {

    private Gson gson = new Gson();
    private Bot myBot;

    public static NetworkingClient connectNew(String[] args, Bot myBot) throws Exception {
        Args parsedArgs = new Args();
        JCommander jCommander = JCommander.newBuilder()
                .addObject(parsedArgs)
                .build();
        jCommander.parse(args);

        // If --help flag is provided, display help
        if (parsedArgs.help) {
            jCommander.setProgramName("my-bot.jar");
            jCommander.usage();
            return null;
        }

        // Setup headers
        Map<String,String> httpHeaders = new HashMap<>();
        httpHeaders.put("token", parsedArgs.token);

        NetworkingClient c = new NetworkingClient(new URI("ws://localhost:" + parsedArgs.port), httpHeaders, myBot);
        // Disable connection lost timeout
        c.setConnectionLostTimeout(0);

        c.connect();

        return c;
    }

    private NetworkingClient(URI serverUri, Map<String, String> httpHeaders, Bot myBot) {
        super(serverUri, httpHeaders);
        this.myBot = myBot;
    }

    @Override
    public void onMessage(ByteBuffer bytes) {}

    @Override
    public void onClose(int code, String reason, boolean remote) {
        System.out.println("Connection closed, exiting...");
        System.exit(0);
    }

    @Override
    public void onError(Exception ex) {
        ex.printStackTrace();
        if (!isOpen()) {
            System.exit(1);
        }
    }

    @Override
    public void onOpen(ServerHandshake handshakeData) {}

    @Override
    public void onMessage(String message) {
        try {
            // Create response object to store commands issued by the bot
            Response response = new Response();

            if (message.contains("\"__type\":\"INITIAL\"")) {
                InitialData data = gson.fromJson(message, InitialData.class);
                response.__uid = data.__uid;
                myBot.setup(data);
            }
            else if (message.contains("\"__type\":\"UPDATE\"")) {
                MatchState data = gson.fromJson(message, MatchState.class);
                response.__uid = data.__uid;
                myBot.update(data, response);
            }
            send(gson.toJson(response));
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: games/planetization/bots/python3/core/networking_client.py
import argparse
import json
import traceback
import websockets

from core.response import Response


async def connect(bot):
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--port", help="Identification token with which bot can connect", required=False)
    parser.add_argument("-t", "--token", help="Port of match generator to which to connect", required=False)
    args = parser.parse_args()

    if args.port is None:
        args.port = "8887"
    if args.token is None:
        args.token = "_"

    async with websockets.connect("ws://localhost:" + args.port, extra_headers={'token': args.token}) as websocket:
        try:
            while True:
                message = await websocket.recv()
                data = json.loads(message)

                response = Response(data['__uid'])
                try:
                    if data['__type'] == 'INITIAL':
                        bot.setup(data)

                    elif data['__type'] == 'UPDATE':
                        bot.update(data, response)

                    response_json = json.dumps(response.__dict__)\
                        .replace("_Response__uid", "__uid")
                    await websocket.send(response_json)
                except Exception as e:
                    print(e)
                    traceback.print_exc()

        except websockets.ConnectionClosed:
            print("Connection closed. Exiting...")
        except Exception:
            traceback.print_exc()
